Test failure in build: BufTest.relativeReadOfUnsignedByteMustNotBoundsCheckWhenReadOffsetAndSizeIsEqualToWriteOffset

https://github.com/netty/netty-incubator-buffer-api/actions/runs/385257419

```
2020-11-26T13:07:26.5820359Z [INFO] -------------------------------------------------------
2020-11-26T13:07:26.5821826Z [INFO]  T E S T S
2020-11-26T13:07:26.5838079Z [INFO] -------------------------------------------------------
2020-11-26T13:07:27.3085876Z [ERROR] WARNING: Using incubator modules: jdk.incubator.foreign
2020-11-26T13:07:31.7609030Z [INFO] Running io.netty.buffer.api.BufTest
2020-11-26T13:08:03.2065216Z [INFO] Running io.netty.buffer.api.BufTest$CleanerTests
2020-11-26T13:08:04.8472333Z [ERROR] Tests run: 33621, Failures: 0, Errors: 1, Skipped: 2, Time elapsed: 0.039 s <<< FAILURE! - in io.netty.buffer.api.BufTest$CleanerTests
2020-11-26T13:08:04.8533154Z [ERROR] io.netty.buffer.api.BufTest.relativeReadOfUnsignedByteMustNotBoundsCheckWhenReadOffsetAndSizeIsEqualToWriteOffset(Fixture)[124]  Time elapsed: 0.013 s  <<< ERROR!
2020-11-26T13:08:04.8538438Z java.lang.IllegalStateException: This segment is already closed
2020-11-26T13:08:04.8541560Z 	at jdk.incubator.foreign/jdk.internal.foreign.AbstractMemorySegmentImpl.checkValidState(AbstractMemorySegmentImpl.java:414)
2020-11-26T13:08:04.8546105Z 	at jdk.incubator.foreign/jdk.internal.foreign.AbstractMemorySegmentImpl.close(AbstractMemorySegmentImpl.java:335)
2020-11-26T13:08:04.8548951Z 	at io.netty.buffer.api.memseg.MemSegBuf.lambda$static$0(MemSegBuf.java:47)
2020-11-26T13:08:04.8551444Z 	at io.netty.buffer.api.SizeClassedMemoryPool.dispose(SizeClassedMemoryPool.java:127)
2020-11-26T13:08:04.8554171Z 	at io.netty.buffer.api.SizeClassedMemoryPool.drop(SizeClassedMemoryPool.java:94)
2020-11-26T13:08:04.8556806Z 	at io.netty.buffer.api.SizeClassedMemoryPool.drop(SizeClassedMemoryPool.java:27)
2020-11-26T13:08:04.8559636Z 	at io.netty.buffer.api.NativeMemoryCleanerDrop.lambda$accept$0(NativeMemoryCleanerDrop.java:71)
2020-11-26T13:08:04.8562437Z 	at java.base/jdk.internal.ref.CleanerImpl$PhantomCleanableRef.performCleanup(CleanerImpl.java:178)
2020-11-26T13:08:04.8564975Z 	at java.base/jdk.internal.ref.PhantomCleanable.clean(PhantomCleanable.java:133)
2020-11-26T13:08:04.8567554Z 	at io.netty.buffer.api.NativeMemoryCleanerDrop$GatedCleanable.clean(NativeMemoryCleanerDrop.java:92)
2020-11-26T13:08:04.8570456Z 	at io.netty.buffer.api.NativeMemoryCleanerDrop.drop(NativeMemoryCleanerDrop.java:47)
2020-11-26T13:08:04.8573279Z 	at io.netty.buffer.api.NativeMemoryCleanerDrop.drop(NativeMemoryCleanerDrop.java:27)
2020-11-26T13:08:04.8575567Z 	at io.netty.buffer.api.RcSupport.close(RcSupport.java:58)
2020-11-26T13:08:04.8577331Z 	at io.netty.buffer.api.CompositeBuf$1.drop(CompositeBuf.java:35)
2020-11-26T13:08:04.8579334Z 	at io.netty.buffer.api.CompositeBuf$1.drop(CompositeBuf.java:31)
2020-11-26T13:08:04.8581094Z 	at io.netty.buffer.api.RcSupport.close(RcSupport.java:58)
2020-11-26T13:08:04.8651523Z 	at io.netty.buffer.api.BufTest.relativeReadOfUnsignedByteMustNotBoundsCheckWhenReadOffsetAndSizeIsEqualToWriteOffset(BufTest.java:1665)
2020-11-26T13:08:04.8657190Z 	at jdk.internal.reflect.GeneratedMethodAccessor181.invoke(Unknown Source)
2020-11-26T13:08:04.8660347Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2020-11-26T13:08:04.8662967Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
2020-11-26T13:08:04.8665388Z 	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:688)
2020-11-26T13:08:04.8668519Z 	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
2020-11-26T13:08:04.8672005Z 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
2020-11-26T13:08:04.8675946Z 	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
2020-11-26T13:08:04.8679818Z 	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
2020-11-26T13:08:04.8768058Z 	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestTemplateMethod(TimeoutExtension.java:92)
2020-11-26T13:08:04.8772447Z 	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
2020-11-26T13:08:04.8775977Z 	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
2020-11-26T13:08:04.8779787Z 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
2020-11-26T13:08:04.8784101Z 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
2020-11-26T13:08:04.8788713Z 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
2020-11-26T13:08:04.8793298Z 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
2020-11-26T13:08:04.8797034Z 	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
2020-11-26T13:08:04.8800279Z 	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
2020-11-26T13:08:04.8805824Z 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:210)
2020-11-26T13:08:04.8810136Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-11-26T13:08:04.8814831Z 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:206)
2020-11-26T13:08:04.8819386Z 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:131)
2020-11-26T13:08:04.8823565Z 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:65)
2020-11-26T13:08:04.8827590Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
2020-11-26T13:08:04.8831655Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-11-26T13:08:04.8835720Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
2020-11-26T13:08:04.8839040Z 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
2020-11-26T13:08:04.8842388Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
2020-11-26T13:08:04.8890204Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-11-26T13:08:04.8894699Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
2020-11-26T13:08:04.8899062Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
2020-11-26T13:08:04.8904620Z 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
2020-11-26T13:08:04.8910518Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask$DefaultDynamicTestExecutor.execute(NodeTestTask.java:212)
2020-11-26T13:08:04.8914522Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask$DefaultDynamicTestExecutor.execute(NodeTestTask.java:192)
2020-11-26T13:08:04.8918725Z 	at org.junit.jupiter.engine.descriptor.TestTemplateTestDescriptor.execute(TestTemplateTestDescriptor.java:139)
2020-11-26T13:08:04.8923277Z 	at org.junit.jupiter.engine.descriptor.TestTemplateTestDescriptor.lambda$execute$2(TestTemplateTestDescriptor.java:107)
2020-11-26T13:08:04.8926447Z 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
2020-11-26T13:08:04.8928459Z 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
2020-11-26T13:08:04.8930624Z 	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
2020-11-26T13:08:04.8932810Z 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
2020-11-26T13:08:04.8934823Z 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
2020-11-26T13:08:04.8936834Z 	at java.base/java.util.stream.ReferencePipeline$15$1.accept(ReferencePipeline.java:539)
2020-11-26T13:08:04.8939017Z 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
2020-11-26T13:08:04.8941177Z 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
2020-11-26T13:08:04.8943338Z 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
2020-11-26T13:08:04.8945355Z 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
2020-11-26T13:08:04.8947359Z 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
2020-11-26T13:08:04.8949580Z 	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
2020-11-26T13:08:04.8951933Z 	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:755)
2020-11-26T13:08:04.8954143Z 	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:274)
2020-11-26T13:08:04.8956309Z 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
2020-11-26T13:08:04.8958484Z 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
2020-11-26T13:08:04.8960824Z 	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
2020-11-26T13:08:04.9016608Z 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
2020-11-26T13:08:04.9019270Z 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
2020-11-26T13:08:04.9021772Z 	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
2020-11-26T13:08:04.9023925Z 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
2020-11-26T13:08:04.9026190Z 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
2020-11-26T13:08:04.9028576Z 	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:594)
2020-11-26T13:08:04.9030883Z 	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:274)
2020-11-26T13:08:04.9033025Z 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
2020-11-26T13:08:04.9035158Z 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
2020-11-26T13:08:04.9037284Z 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
2020-11-26T13:08:04.9039500Z 	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
2020-11-26T13:08:04.9042119Z 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
2020-11-26T13:08:04.9044990Z 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
2020-11-26T13:08:04.9047476Z 	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
2020-11-26T13:08:04.9049617Z 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
2020-11-26T13:08:04.9051861Z 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
2020-11-26T13:08:04.9054267Z 	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:594)
2020-11-26T13:08:04.9056542Z 	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:274)
2020-11-26T13:08:04.9058754Z 	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
2020-11-26T13:08:04.9061113Z 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
2020-11-26T13:08:04.9063653Z 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
2020-11-26T13:08:04.9066112Z 	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
2020-11-26T13:08:04.9068273Z 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
2020-11-26T13:08:04.9070507Z 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
2020-11-26T13:08:04.9072891Z 	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:594)
2020-11-26T13:08:04.9076319Z 	at org.junit.jupiter.engine.descriptor.TestTemplateTestDescriptor.execute(TestTemplateTestDescriptor.java:107)
2020-11-26T13:08:04.9080754Z 	at org.junit.jupiter.engine.descriptor.TestTemplateTestDescriptor.execute(TestTemplateTestDescriptor.java:42)
2020-11-26T13:08:04.9125200Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
2020-11-26T13:08:04.9142591Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-11-26T13:08:04.9154332Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
2020-11-26T13:08:04.9165254Z 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
2020-11-26T13:08:04.9168714Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
2020-11-26T13:08:04.9172780Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-11-26T13:08:04.9177438Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
2020-11-26T13:08:04.9181546Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
2020-11-26T13:08:04.9184150Z 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
2020-11-26T13:08:04.9188833Z 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
2020-11-26T13:08:04.9194753Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
2020-11-26T13:08:04.9198826Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-11-26T13:08:04.9203088Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
2020-11-26T13:08:04.9226776Z 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
2020-11-26T13:08:04.9304204Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
2020-11-26T13:08:04.9318668Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-11-26T13:08:04.9390975Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
2020-11-26T13:08:04.9439162Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
2020-11-26T13:08:04.9451607Z 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
2020-11-26T13:08:04.9457060Z 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
2020-11-26T13:08:04.9462860Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
2020-11-26T13:08:04.9467012Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-11-26T13:08:04.9471078Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
2020-11-26T13:08:04.9474449Z 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
2020-11-26T13:08:04.9477796Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
2020-11-26T13:08:04.9481880Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-11-26T13:08:04.9486657Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
2020-11-26T13:08:04.9490712Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
2020-11-26T13:08:04.9496287Z 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
2020-11-26T13:08:04.9502642Z 	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
2020-11-26T13:08:04.9507719Z 	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
2020-11-26T13:08:04.9511633Z 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:220)
2020-11-26T13:08:04.9514636Z 	at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$6(DefaultLauncher.java:188)
2020-11-26T13:08:04.9518146Z 	at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:202)
2020-11-26T13:08:04.9521675Z 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:181)
2020-11-26T13:08:04.9537448Z 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:128)
2020-11-26T13:08:04.9541930Z 	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:188)
2020-11-26T13:08:04.9546392Z 	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:154)
2020-11-26T13:08:04.9550894Z 	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:128)
2020-11-26T13:08:04.9554666Z 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
2020-11-26T13:08:04.9557850Z 	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
2020-11-26T13:08:04.9560474Z 	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
2020-11-26T13:08:04.9563326Z 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)
2020-11-26T13:08:04.9564685Z 
2020-11-26T13:08:05.0261296Z [INFO] Tests run: 0, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 31.633 s - in io.netty.buffer.api.BufTest
2020-11-26T13:08:05.0271675Z [INFO] Running io.netty.buffer.api.ScopeTest
2020-11-26T13:08:05.0380511Z [INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.077 s - in io.netty.buffer.api.ScopeTest
2020-11-26T13:08:06.4946990Z [INFO] 
2020-11-26T13:08:06.4965073Z [INFO] Results:
2020-11-26T13:08:06.4974725Z [INFO] 
2020-11-26T13:08:06.4995646Z [ERROR] Errors: 
2020-11-26T13:08:06.5023674Z [ERROR]   BufTest.relativeReadOfUnsignedByteMustNotBoundsCheckWhenReadOffsetAndSizeIsEqualToWriteOffset:1665 ? IllegalState
2020-11-26T13:08:06.5028173Z [INFO] 
2020-11-26T13:08:06.5029154Z [ERROR] Tests run: 33621, Failures: 0, Errors: 1, Skipped: 1
```

```
2020-11-30T07:28:53.1319422Z [INFO] -------------------------------------------------------
2020-11-30T07:28:53.1320441Z [INFO]  T E S T S
2020-11-30T07:28:53.1321311Z [INFO] -------------------------------------------------------
2020-11-30T07:28:54.1803477Z [ERROR] WARNING: Using incubator modules: jdk.incubator.foreign
2020-11-30T07:28:57.9366361Z [INFO] Running io.netty.buffer.api.BufTest
2020-11-30T07:29:30.1458267Z [INFO] Running io.netty.buffer.api.BufTest$CleanerTests
2020-11-30T07:29:32.0566747Z [ERROR] Tests run: 33621, Failures: 0, Errors: 1, Skipped: 2, Time elapsed: 0.031 s <<< FAILURE! - in io.netty.buffer.api.BufTest$CleanerTests
2020-11-30T07:29:32.0622600Z [ERROR] io.netty.buffer.api.BufTest.offsettedGetOfDoubleMustBoundsCheckWhenReadOffsetIsGreaterThanWriteOffset(Fixture)[106]  Time elapsed: 0.035 s  <<< ERROR!
2020-11-30T07:29:32.0642085Z java.lang.IllegalStateException: This segment is already closed
2020-11-30T07:29:32.0644573Z 	at jdk.incubator.foreign/jdk.internal.foreign.AbstractMemorySegmentImpl.checkValidState(AbstractMemorySegmentImpl.java:414)
2020-11-30T07:29:32.0652972Z 	at jdk.incubator.foreign/jdk.internal.foreign.AbstractMemorySegmentImpl.close(AbstractMemorySegmentImpl.java:335)
2020-11-30T07:29:32.0656292Z 	at io.netty.buffer.api.memseg.MemSegBuf.lambda$static$0(MemSegBuf.java:47)
2020-11-30T07:29:32.0678191Z 	at io.netty.buffer.api.SizeClassedMemoryPool.dispose(SizeClassedMemoryPool.java:127)
2020-11-30T07:29:32.0680833Z 	at io.netty.buffer.api.SizeClassedMemoryPool.drop(SizeClassedMemoryPool.java:94)
2020-11-30T07:29:32.0684140Z 	at io.netty.buffer.api.SizeClassedMemoryPool.drop(SizeClassedMemoryPool.java:27)
2020-11-30T07:29:32.0687405Z 	at io.netty.buffer.api.NativeMemoryCleanerDrop.lambda$accept$0(NativeMemoryCleanerDrop.java:71)
2020-11-30T07:29:32.0690504Z 	at java.base/jdk.internal.ref.CleanerImpl$PhantomCleanableRef.performCleanup(CleanerImpl.java:178)
2020-11-30T07:29:32.0693239Z 	at java.base/jdk.internal.ref.PhantomCleanable.clean(PhantomCleanable.java:133)
2020-11-30T07:29:32.0697040Z 	at io.netty.buffer.api.NativeMemoryCleanerDrop$GatedCleanable.clean(NativeMemoryCleanerDrop.java:92)
2020-11-30T07:29:32.0705884Z 	at io.netty.buffer.api.NativeMemoryCleanerDrop.drop(NativeMemoryCleanerDrop.java:47)
2020-11-30T07:29:32.0761175Z 	at io.netty.buffer.api.NativeMemoryCleanerDrop.drop(NativeMemoryCleanerDrop.java:27)
2020-11-30T07:29:32.0763640Z 	at io.netty.buffer.api.RcSupport.close(RcSupport.java:58)
2020-11-30T07:29:32.0765352Z 	at io.netty.buffer.api.CompositeBuf$1.drop(CompositeBuf.java:35)
2020-11-30T07:29:32.0767243Z 	at io.netty.buffer.api.CompositeBuf$1.drop(CompositeBuf.java:31)
2020-11-30T07:29:32.0769553Z 	at io.netty.buffer.api.RcSupport.close(RcSupport.java:58)
2020-11-30T07:29:32.0774818Z 	at io.netty.buffer.api.BufTest.offsettedGetOfDoubleMustBoundsCheckWhenReadOffsetIsGreaterThanWriteOffset(BufTest.java:3706)
```

--- a/src/main/java/io/netty/buffer/api/NativeMemoryCleanerDrop.java
+++ b/src/main/java/io/netty/buffer/api/NativeMemoryCleanerDrop.java
@@ -61,9 +61,9 @@ class NativeMemoryCleanerDrop implements Drop<Buf> {
         var mem = manager.unwrapRecoverableMemory(buf);
         var delegate = this.delegate;
         WeakReference<Buf> ref = new WeakReference<>(buf);
-        AtomicBoolean gate = new AtomicBoolean();
+        AtomicBoolean gate = new AtomicBoolean(true);
         cleanable = new GatedCleanable(gate, CLEANER.register(this, () -> {
-            if (gate.compareAndSet(false, true)) {
+            if (gate.getAndSet(false)) {
                 Buf b = ref.get();
                 if (b == null) {
                     pool.recoverMemory(mem);
@@ -84,7 +84,7 @@ class NativeMemoryCleanerDrop implements Drop<Buf> {
         }
 
         public void disable() {
-            gate.set(true);
+            gate.set(false);
         }
 
         @Override

--- a/src/main/java/io/netty/buffer/api/SizeClassedMemoryPool.java
+++ b/src/main/java/io/netty/buffer/api/SizeClassedMemoryPool.java
@@ -70,7 +70,7 @@ class SizeClassedMemoryPool implements Allocator, AllocatorControl, Drop<Buf> {
                 Send<Buf> send;
                 while ((send = v.poll()) != null) {
                     try {
-                        dispose(send.receive());
+                        send.receive().close();
                     } catch (Exception e) {
                         capturedExceptions.add(e);
                     }
@@ -86,12 +86,16 @@ class SizeClassedMemoryPool implements Allocator, AllocatorControl, Drop<Buf> {
 
     @Override
     public void drop(Buf buf) {
+        if (closed) {
+            dispose(buf);
+            return;
+        }
         var sizeClassPool = getSizeClassPool(buf.capacity());
         sizeClassPool.offer(buf.send());
         if (closed) {
             Send<Buf> send;
             while ((send = sizeClassPool.poll()) != null) {
-                dispose(send.receive());
+                send.receive().close();
             }
         }
     }

